expression_evaluate.go

```
@@ -33,13 +34,167 @@ func (c evaluateContext) evaluate(node ast.Node, data any) (any, error) {
+	case ast.LessThan:
+		return a < b, nil
+	case ast.GreaterThanEquals:
+		return a >= b, nil
+	case ast.LessThanEquals:
+		return a <= b, nil
+	case ast.And, ast.Or:
+		return nil, fmt.Errorf("attempted logical operation %s on numeric input %T", op, a)
+	default:
+		return nil, fmt.Errorf("numeric eval missing case for logical operation %s", op)
+	}
+}
+func evalBooleanOperation(a, b bool, op ast.MathOperationType) (any, error) {
+	switch op {
+	case ast.Invalid:
+		return nil, fmt.Errorf("attempted to perform invalid operation")
+	case ast.Power, ast.Modulus, ast.Divide, ast.Multiply, ast.Subtract, ast.Add:
+		return nil, fmt.Errorf("attempted to perform math operation '%s' on boolean", op)
+	case ast.GreaterThan, ast.LessThan, ast.GreaterThanEquals, ast.LessThanEquals:
+		return nil, fmt.Errorf("attempted to perform invalid operation '%s' on boolean", op)
+	case ast.Equals:
+		return a == b, nil
+	case ast.NotEquals:
+		return a != b, nil
+	case ast.And:
+		return a && b, nil
+	case ast.Or:
+		return a || b, nil
+	default:
+		return nil, fmt.Errorf("numeric eval missing case for logical operation %s", op)
+	}
+}
+
+func evalStringOperation(a, b string, op ast.MathOperationType) (any, error) {
+	switch op {
+	case ast.Invalid:
+		return nil, fmt.Errorf("attempted to perform invalid operation")
+	case ast.Add:
+		// Concatenate
+		return a + b, nil
+	case ast.Subtract, ast.Multiply, ast.Divide, ast.Modulus, ast.Power:
+		return nil, fmt.Errorf("strings do not support operator '%s'", op)
+	case ast.Equals:
+		return a == b, nil
+	case ast.NotEquals:
+		return a != b, nil
+	case ast.GreaterThan:
+		return a > b, nil
+	case ast.LessThan:
+		return a < b, nil
+	case ast.GreaterThanEquals:
+		return a >= b, nil
+	case ast.LessThanEquals:
+		return a <= b, nil
+	case ast.And, ast.Or:
+		return nil, fmt.Errorf("attempted logical operation %s on string input", op)
+	default:
+		return nil, fmt.Errorf("string eval missing case for logical operation %s", op)
```

This sounds more like an assert/panic rather than returning a user error?

---

Sure. Seems aggressive, but changed.

---

Well, it doesn't sound like we should have been able to get here without an internal logic error. If that's not the case, then my comment may have been misguided. Returning a user error makes it sound like the user did something wrong and should fix their workflow expression. If they can't do that, because it's not their problem, a panic is probably more appropriate because they'll know to report an ARCA problem rather than spending a day trying to fix their expression.